experimental/trusted_shuffler/trusted_shuffler/src/tests.rs

```diff
@@ -0,0 +1,123 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::TrustedShuffler;
+use futures::future::join_all;
+use std::{sync::Arc, time::Duration};
+use tokio::time::timeout;
+
+const TEST_ANONYMITY_VALUE: usize = 10;
+const TEST_TIMEOUT_MILLISECONDS: u64 = 100;
+
+// A test function that processes requests.
+// In the real use-case, this function should send requests to the backend and
+// return responses.
+async fn handle_request(request: Vec<u8>) -> Vec<u8> {
+    generate_response(request)
+}
+
+// Non-async version of the `handle_request` function used to create expected responses.
+fn generate_response(request: Vec<u8>) -> Vec<u8> {
+    let parsed_request = String::from_utf8(request).expect("Couldn't parse request");
+    format!("Response for: {}", parsed_request)
+        .to_string()
+        .as_bytes()
+        .to_vec()
+}
+
+// Generates a request and a corresponding response from a string.
+fn generate_request_response(data: &str) -> (Vec<u8>, Vec<u8>) {
+    let request = format!("Request: {}", data).to_string().as_bytes().to_vec();
+    let response = generate_response(request.clone());
+    (request, response)
+}
+
+#[tokio::test]
+async fn trusted_shuffler_test() {
+    let anonymity_value = 2;
+    let trusted_shuffler = Arc::new(TrustedShuffler::new(anonymity_value, handle_request));
+
+    let (request, expected_response) = generate_request_response("Test");
+    let (background_request, expected_background_response) =
+        generate_request_response("Background test");
+
+    let trusted_shuffler_clone = trusted_shuffler.clone();
+    let background_request_clone = background_request.clone();
+    let background_result = tokio::spawn(async move {
+        trusted_shuffler_clone
+            .invoke(background_request_clone)
+            .await
+    });
+
+    let response = trusted_shuffler.invoke(request.clone()).await;
+    assert!(response.is_ok());
+    assert_eq!(expected_response, response.unwrap());
+
+    let background_result = background_result.await;
+    assert!(background_result.is_ok());
+    let background_response = background_result.unwrap();
+    assert!(background_response.is_ok());
+    assert_eq!(expected_background_response, background_response.unwrap());
+}
+
+#[tokio::test]
+async fn trusted_shuffler_waiting_test() {
```

Comment:
What am I missing... we are not doing anything with timing yet, are we? What are we testing here?

Reply:
ah, I didn't realize! :) Can we rename the test or comment that?